_posts/2016-04-07-developing-nectar-extensions-part-1.md

```diff
@@ -0,0 +1,168 @@
+
+And add to install.ex the calls:
+
+<script src="https://gist.github.com/nimish-mehta/db7883f628837e7ebca5a1945c4d1bfe.js"></script>
+
+Now we can see the added routes from Nectar:
+
+<script src="https://gist.github.com/nimish-mehta/994e51defad0787eb88e6611219066fb.js?file=route.bash"></script>
+
+So far so good, we have modified and added routes and controller to Nectar's router. Time to see our handiwork in action, start the server from Nectar application with:
+
+<script src="https://gist.github.com/nimish-mehta/994e51defad0787eb88e6611219066fb.js?file=server.bash"></script>
+
+And, visit 127.0.0.1:4000/favorite and click on mark to like a product.
+
+![Missing Layout](assets/images/before_layout.png){: .center-image }
+
+But things don't seem right do they, our Nectar layout has been replaced with the default one used by phoenix. Let's rectify that.
+
+Update layout_view.ex as:
+
+<script src="https://gist.github.com/nimish-mehta/ceb97b1c0539f94d2a4bbf95b202a861.js"></script>
+
+and recompile and restart the server
+
+<script src="https://gist.github.com/nimish-mehta/994e51defad0787eb88e6611219066fb.js?file=compile.bash"></script>
+
+On our next visit:
+
+![Layout Present](assets/images/after_layout.png){: .center-image }
+
+Much better.
+
+> __Note__: When we need to change the extension code while running the server we will have to recompile and reload. We don't have anything in Nectar right now to monitor all extensions file and do an automatic compilation and code reload.
+
+
+#Testing#
+We are almost done now. To ensure that we know when things break we should consider adding a few test cases. For that we need to make sure that Nectar migrations are run before running the migrations for favorite products and we need the Nectar repo running as well.
+
+For the former we could update the test_helper.ex with:
+
+<script src="https://gist.github.com/nimish-mehta/795a1eacd54f876f774d3d91abcc8fb3.js"></script>
+
+But things are not so smooth for accessing the Nectar Repo. Which brings us to what we think is the ultimate downfall of this approach:
+
+### An Untestable soution ###
+
+Ideally, running ```mix test``` should work and we should see our test running green, unfortunately this requires Nectar to be compiled before running the tests, which is impossible since Nectar depends upon the extension\_manager to be compiled which depends upon all the extensions to be compiled, resulting in a cyclic dependency. Also we used Nectar's repo for all database work. That works because we were running our server through Nectar and the repo was started in Nectar's Supervision tree. Which again adds an implicit requirement Nectar application is available and ready to be started during test time. We could replace ```Nectar.Repo``` with ```FavoriteProducts.Repo``` using compile time conditionals i.e. when ```MIX_ENV==test```, but it is a can of worms we would rather avoid right now.
```

Comment:
That works because we were running our server through Nectar and the repo was started in Nectar's Supervision tree**, **which again adds an implicit requirement **i.e **Nectar application is available and ready to be started during test time.